File: Examples/OxideRegrowth/OxideRegrowth.py
DIM = 3

if DIM is 3:
    stabFac = 0.145
    import viennaps3d as vps
else:
    stabFac = 0.245
    import viennaps2d as vps

vps.Logger.setLogLevel(vps.LogLevel.INTERMEDIATE)

# parse the parameters
params = vps.ReadConfigFile("config.txt")

stability = (
    2
    * params["diffusionCoefficient"]
    / max(params["scallopVelocity"], params["centerVelocity"])
)
print(f"Stability: {stability}")
if 0.5 * stability <= params["gridDelta"]:
    print("Unstable parameters. Reduce grid spacing!")

domain = vps.Domain()
vps.MakeStack(
    domain=domain,
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=0.0,
    numLayers=int(params["numLayers"]),
    layerHeight=params["layerHeight"],
    substrateHeight=params["substrateHeight"],
    holeRadius=0.0,
    trenchWidth=params["trenchWidth"],
    maskHeight=0.0,
).apply()
# copy top layer for deposition
domain.duplicateTopLevelSet(vps.Material.Polymer)

domain.generateCellSet(
    params["substrateHeight"] + params["numLayers"] * params["layerHeight"] + 10.0, True
)
cellSet = domain.getCellSet()
cellSet.addScalarData("byproductSum", 0.0)
cellSet.writeVTU("initial.vtu")
# we need neighborhood information for solving the
# convection-diffusion equation on the cell set
cellSet.buildNeighborhood()

# The redeposition model captures byproducts from the selective etching
# process in the cell set. The byproducts are then distributed by solving a
# convection-diffusion equation on the cell set.
model = vps.OxideRegrowthModel(
    nitrideEtchRate=params["nitrideEtchRate"] / 60.0,
    oxideEtchRate=params["oxideEtchRate"] / 60.0,
    redepositionRate=params["redepositionRate"],
    redepositionThreshold=params["redepositionThreshold"],
    redepositionTimeInt=params["redepositionTimeInt"],
    diffusionCoefficient=params["diffusionCoefficient"],
    sinkStrength=params["sink"],
    scallopVelocity=params["scallopVelocity"],
    centerVelocity=params["centerVelocity"],
    topHeight=params["substrateHeight"] + params["numLayers"] * params["layerHeight"],
    centerWidth=params["trenchWidth"],
    stabilityFactor=stabFac,
)

process = vps.Process()
process.setDomain(domain)
process.setProcessModel(model)
process.setProcessDuration(params["targetEtchDepth"] / params["nitrideEtchRate"] * 60.0)
process.setPrintTimeInterval(30.0)

process.apply()

vps.WriteVisualizationMesh(domain, "FinalStack").apply()

File: examples/oxideRegrowth/oxideRegrowth.cpp
#include <geometries/psMakeStack.hpp>
#include <psProcess.hpp>

#include <models/psOxideRegrowth.hpp>

namespace ps = viennaps;

int main(int argc, char **argv) {
  using NumericType = double;

  omp_set_num_threads(12);
  constexpr int D = 2;

  ps::Logger::setLogLevel(ps::LogLevel::INTERMEDIATE);

  // Parse the parameters
  ps::util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
    return 1;
  }

  const NumericType stability =
      2 * params.get("diffusionCoefficient") /
      std::max(params.get("centerVelocity"), params.get("scallopVelocity"));
  std::cout << "Stability: " << stability << std::endl;
  constexpr NumericType timeStabilityFactor = D == 2 ? 0.245 : 0.145;
  if (0.5 * stability <= params.get("gridDelta")) {
    std::cout << "Unstable parameters. Reduce grid spacing!" << std::endl;
    return -1;
  }

  auto domain = ps::SmartPointer<ps::Domain<NumericType, D>>::New(
      params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"));
  ps::MakeStack<NumericType, D>(
      domain, params.get<int>("numLayers"), params.get("layerHeight"),
      params.get("substrateHeight"), 0. /*holeRadius*/,
      params.get("trenchWidth"), 0. /*maskHeight*/, 0. /*taperAngle*/)
      .apply();
  // copy top layer for deposition
  domain->duplicateTopLevelSet(ps::Material::Polymer);

  domain->generateCellSet(
      params.get("substrateHeight") +
          params.get("numLayers") * params.get("layerHeight") + 10.,
      ps::Material::GAS, true /* true means cell set above surface */);
  auto &cellSet = domain->getCellSet();
  cellSet->addScalarData("byproductSum", 0.);
  cellSet->writeVTU("initial.vtu");
  if constexpr (D == 3) {
    std::array<bool, D> boundaryConds = {false};
    boundaryConds[1] = true;
    cellSet->setPeriodicBoundary(boundaryConds);
  }
  // we need neighborhood information for solving the
  // convection-diffusion equation on the cell set
  cellSet->buildNeighborhood();

  // The redeposition model captures byproducts from the selective etching
  // process in the cell set. The byproducts are then distributed by solving a
  // convection-diffusion equation on the cell set.
  auto model = ps::SmartPointer<ps::OxideRegrowth<NumericType, D>>::New(
      params.get("nitrideEtchRate") / 60., params.get("oxideEtchRate") / 60.,
      params.get("redepositionRate"), params.get("redepositionThreshold"),
      params.get("redepositionTimeInt"), params.get("diffusionCoefficient"),
      params.get("sink"), params.get("scallopVelocity"),
      params.get("centerVelocity"),
      params.get("substrateHeight") +
          params.get("numLayers") * params.get("layerHeight"),
      params.get("trenchWidth"), timeStabilityFactor);

  ps::Process<NumericType, D> process;
  process.setDomain(domain);
  process.setProcessModel(model);
  process.setProcessDuration(params.get("targetEtchDepth") /
                             params.get("nitrideEtchRate") * 60.);
  process.apply();

  domain->saveVolumeMesh("finalStack");
}
